### PR TITLE
[ebpf/manager] Resolve ELF symbol from memory (/proc/pid/maps) if the library file is not accessible/available

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

Resolve ELF symbol from memory (/proc/pid/maps) if the library file is not accessible/available

### Motivation

FindSymbolOffsets() will try to extract symbols from the library ELF file.
If it can't fallback to parse the ELF available from memory (proc/pid/maps)
But the original ELF sections could not be avaiable (not mapped in memory)
so the last chance is to get the table file offset (DYNTAB, SYMTAB)
from the prog ELF section DT_DYNAMIC and reconstruct ELF section
to extract dynamic symbols.

OpenAndListSymbols(path) : read the ELF thanks to debug/elf

OpenAndListSymbolsFromPID(pid, path) : read the ELF thanks to elfmap

If the .so is not accessible, elfmap would ptrace() PEEK_DATA and copy the mapped lib.
If the caller lookup to himself, so the same memspace, we just memcpy the sections.

### CODE REVIEW

**Comment by brycekahle** [comment]
@nplanel Is this an exact copy of what is in Datadog/ebpf or are there changes?

**Comment by nplanel** [comment]
> @nplanel Is this an exact copy of what is in Datadog/ebpf or are there changes?

That's a copy, no changes

File: utils_test.go [hunk 1]
```diff
@@ -31,3 +34,77 @@ func TestGenerateEventName(t *testing.T) {
...
+	}
+	defer f.Close()
+	return sym
+}
+
+func TestOpenAndListSymbols(t *testing.T) {
+	path := detectLDLoaderPath(t)
+
+	syms := openAndListSymbols(t, path)
+	t.Log("NR symbol loaded:", len(syms))
```
Comment (brycekahle) [inline utils_test.go:69]: Can we add some asserts to this test?

File: utils_test.go [hunk 2]
```diff
@@ -31,3 +34,77 @@ func TestGenerateEventName(t *testing.T) {
...
+}
+
+func TestOpenAndListSymbolsFromPID(t *testing.T) {
+	path := detectLDLoaderPath(t)
+
+	syms := openAndListSymbolsFromPID(t, path)
+	t.Log("NR symbol loaded:", len(syms))
```
Comment (brycekahle) [inline utils_test.go:76]: Can we add some asserts to this test?

File: utils.go [hunk 3]
```diff
@@ -405,11 +550,40 @@ func SanitizeUprobeAddresses(f *elf.File, syms []elf.Symbol) {
...
+					continue
+				}
+				f, syms, err = OpenAndListSymbolsFromPID(pid, path)
+				if err == nil {
+					found = true
+					goto found_syms
+				}
+			}
+		}
+		if !found {
```
Comment (brycekahle) [inline utils.go:580]: `found` will always be `false` here, because as soon as it is set to `true`, you use a `goto`.

File: utils.go [hunk 4]
```diff
@@ -405,11 +550,40 @@ func SanitizeUprobeAddresses(f *elf.File, syms []elf.Symbol) {
...
+				pidStr := filepath.Base(pidPath)
+				var pid int
+				pid, err = strconv.Atoi(pidStr)
+				if err != nil {
+					continue
+				}
+				f, syms, err = OpenAndListSymbolsFromPID(pid, path)
+				if err == nil {
+					found = true
+					goto found_syms
```
Comment (brycekahle) [inline utils.go:576]: If we can avoid a `goto` we should, as it makes code much harder to read and reason about. I think using a `break` with a label would be easier to understand.

File: utils_test.go (hunk 1 above)
Comment (nplanel) [reply utils_test.go:69]: wasn't possible to failed (ie t.Fatal() in the caller) but I move to a more coventional testing/assert using stretchr/testify.NoError()

File: utils.go (hunk 4 above)
Comment (nplanel) [reply utils.go:576]: break with label, I found that counter intuitive (C perspective) as the label need to be declared before the loop, but meanwhile it's a break not a goto (Go perspective)

File: utils.go [hunk 5]
```diff
@@ -15,8 +15,13 @@ import (
 	"strings"
 	"sync"
 	"syscall"
+	"unsafe"
 
 	"golang.org/x/sys/unix"
+
+	"github.com/DataDog/ebpf-manager/elfmap"
```
Comment (r1viollet) [inline utils.go:22]: Is there a separate branch that delivers the `elfmap` package ? I was not able to find it here.

Comment (nplanel) [reply utils.go:23]: @r1viollet sorry didn't see your msg, here the elfmap package

Comment (r1viollet) [reply utils.go:23]: No problem, if you resume this, we also did a prototype at one point here : https://github.com/Gui774ume/utrace/pull/12/files
I'm  not sure if that can be useful.

**Comment by brycekahle** [comment]
I'm going to close this for now, since it has been sitting stale for quite awhile. If you resume this, feel free to open a new PR.

**Pull request closed without merging**